### PR TITLE
 Add helper base class for generating new replication endpoints

### PR DESCRIPTION
**Pull request by erikjohnston**
This will hopefully reduce the boiler plate required to implement new
internal HTTP requests.

### CODE REVIEW

**Comment by erikjohnston** [comment]
> I'm not entirely happy with the cleverness going on here, because it will make clicking through to implementations, type inference, etc not work properly.

Yeah, I'm not entirely happy about it either, but I think its the step in the right direction in terms of avoiding explosions of custom implemented replication endpoints.

I had a quick play around trying to get some better click through/docs in vscode, but haven't had much luck (it doesn't seem to like inherited staticmethods....)

File: synapse/replication/http/_base.py [hunk 1]
```diff
@@ -190,7 +190,9 @@ def register(self, http_server):
         http_server.register_paths(method, [pattern], handler)
 
     def _cached_handler(self, request, txn_id, **kwargs):
-        """Wraps `_handle_request` the responses should be cached.
+        """Called on new incoming requests when caching is enabled. Checks
+        if their is a cached response for the request and returns that,
```
Comment (richvdh) [inline synapse/replication/http/_base.py]: there

File: synapse/replication/http/_base.py [hunk 2]
```diff
@@ -174,8 +185,7 @@ def register(self, http_server):
         url_args = list(self.PATH_ARGS)
         method = "GET"
```
Comment (richvdh) [inline synapse/replication/http/_base.py]: this is dead

File: synapse/replication/http/_base.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        pattern = re.compile("^/_synapse/replication/%s/%s$" % (
+            self.NAME,
+            args
+        ))
+
+        http_server.register_paths(method, [pattern], handler)
+
+    def _cached_handler(self, request, txn_id, **kwargs):
+        """Called on new incoming requests when caching is enabled. Checks
+        if their is a cached response for the request and returns that,
```
Comment (richvdh) [inline synapse/replication/http/_base.py]: their

**Pull request merged**